### PR TITLE
Combine Regex and Pattern

### PR DESCRIPTION
**Pull request by GregoryTravis**
### Pull Request Description

Merge Pattern into Regex.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: distribution/lib/Standard/Examples/0.0.0-dev/src/Main.enso [hunk 1]
```diff
@@ -1,6 +1,7 @@
 from Standard.Base import all
 
 import Standard.Base.Data.Text.Regex.Match.Match
+import Standard.Base.Data.Text.Regex.Regex
```
Comment (jdunkerley) [inline distribution/lib/Standard/Examples/0.0.0-dev/src/Main.enso]: Not needed as in `from Standard.Base import all`

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Parse_To_Table.enso [hunk 2]
```diff
@@ -1,5 +1,7 @@
 from Standard.Base import all
 
+import Standard.Base.Data.Text.Regex.Regex
```
Comment (jdunkerley) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Parse_To_Table.enso]: Not needed as in `from Standard.Base import all`.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Split_Tokenize.enso [hunk 3]
```diff
@@ -1,11 +1,12 @@
 from Standard.Base import all
 
+import Standard.Base.Data.Text.Regex.Regex
```
Comment (jdunkerley) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Split_Tokenize.enso]: Not needed as in `from Standard.Base import all`.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Table_Helpers.enso [hunk 4]
```diff
@@ -1,4 +1,5 @@
 from Standard.Base import all
+import Standard.Base.Data.Text.Regex.Regex
```
Comment (jdunkerley) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Table_Helpers.enso]: Not needed as in `from Standard.Base import all`.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso [hunk 5]
```diff
@@ -1,59 +1,476 @@
 import project.Any.Any
+import project.Errors.Common.Syntax_Error
+import project.Data.Filter_Condition.Filter_Condition
+import project.Data.Map.Map
 import project.Data.Numbers.Integer
+from project.Data.Range.Extensions import all
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso]: can we put these with the other `from` statements. I know we don't have a fixed order but I think imports first then forms in alphabetical order.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso [hunk 6]
```diff
@@ -1,59 +1,476 @@
...
+polyglot java import org.enso.base.Replacer_Cache
+polyglot java import org.enso.base.Text_Utils
 polyglot java import org.enso.base.Regex_Utils
 
-## Compile the provided `expression` into a regex pattern that can be used for
-   matching.
+type Regex
+    ## internal_regex_object : RegexObject (Truffle)
+       (See https://github.com/oracle/graal/blob/master/regex/docs/README.md)
+    Value (internal_regex_object : Any)
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso]: Should be PRIVATE.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso [hunk 7]
```diff
@@ -1,59 +1,476 @@
...
+         insensitive matching behaves as if it normalises the case of all input
+         text before matching on it.
+
+       If an empty regex is used, `compile` throws an Illegal_Argument error.
+
+       ? Why Compile?
+         While many regex engines are able to cache ad-hoc patterns, it is often
+         useful to be able to manually retain a pattern that you have computed. This
+         function exists so you can hold onto the resultant `Regex` object,
+         instead of immediately proceeding to match using it.
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso]: Don't think we need this comment.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso [hunk 8]
```diff
@@ -1,59 +1,476 @@
...
+        map = polyglot_map_to_map self.internal_regex_object.groups
+        map.keys
+
+    ## Return a map from group number to group name. Only includes named groups.
+    group_nums_to_names : Map Integer Text
+    group_nums_to_names self =
+        map = polyglot_map_to_map self.internal_regex_object.groups
+        map.transform k-> v-> [v, k]
+
+    ## Compile the provided `expression` into a regex pattern that can be used for
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso]:
```suggestion
    ## Compile the provided `expression` into a `Regex` that can be used for
```

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso [hunk 9]
```diff
@@ -1,59 +1,476 @@
...
+
+       If an empty regex is used, `compile` throws an Illegal_Argument error.
+
+       ? Why Compile?
+         While many regex engines are able to cache ad-hoc patterns, it is often
+         useful to be able to manually retain a pattern that you have computed. This
+         function exists so you can hold onto the resultant `Regex` object,
+         instead of immediately proceeding to match using it.
+    compile : Text -> Boolean | Nothing -> Regex ! Regex_Syntax_Error | Illegal_Argument
+    compile expression case_insensitive=Nothing =
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso]: I'd put the statics at the top of the type. Constructor method nexts to actual Constructor.

**Review by jdunkerley — APPROVED** [review]
Looks good but a little comment fixing and shuffling please.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso [hunk 10]
```diff
@@ -1,59 +1,476 @@
...
+    character graphemes, thereby splitting them apart.
+
+    At each step, it yields a Match_Iterator_Value, whivch has either a filler
+    and a match, or just the final filler. A Match_Iterator_Value.Last value is
+    return at the end, and only at the end.
+
+    Optionally, you can call `early_exit` to have it return the remainder of
+    the string, unmatched, as a single Last value. (Used for `replace` with
+    `only_first=True`.)
+type Match_Iterator
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso]: I'm wondering if we should move types like this to `Internal`.

At some point definitely will. May be worth to do now, but not necessary - up to you.

**Review by radeusgd — APPROVED** [review]

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso (hunk 10 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso]: Done

File: distribution/lib/Standard/Examples/0.0.0-dev/src/Main.enso (hunk 1 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Examples/0.0.0-dev/src/Main.enso]: Done

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Parse_To_Table.enso (hunk 2 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Parse_To_Table.enso]: Done

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Split_Tokenize.enso (hunk 3 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Split_Tokenize.enso]: Done

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Table_Helpers.enso (hunk 4 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Table_Helpers.enso]: Done

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso (hunk 5 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso]: Done

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso (hunk 6 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso]: Done

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso (hunk 7 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso]: Done

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso (hunk 8 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso]: Done

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso (hunk 9 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex.enso]: Put `compile` at the top; left `escape` because it's not a constructor method.

**Pull request merged**